# Personalization Homework 1 by Yuan Cheng

Import Libraries

In [4]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from scipy.optimize import minimize

Setup

In [5]:
df = pd.read_csv("winequality-red.csv", sep = ";")

df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [6]:
y = df.pop("quality")
X = df
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2)

Linear Model Equation

In [45]:
def linear_regression(X, beta):
    predictions = list()
    for i in range(len(X)):
        y_hat = sum(x * y for x, y in zip(df.loc[i],beta))
        predictions.append(y_hat)
    return predictions

RSS Equation

In [46]:
def RSS(beta,X,y):
    y_hat = linear_regression(X, beta)
    return np.sum((np.array(y)-np.array(y_hat))**2).astype(float).item()

Optimizing the Model

In [13]:
beta0 = np.random.normal(0,1,X_train.shape[1])
res = minimize(fun=RSS, x0=beta0, args=(X_train, y_train))
beta_hat = res.x
beta_hat

array([-2.10680202e-02, -9.63407572e-02,  8.42118094e-04,  1.48925666e-03,
        2.10796400e-02,  5.78577211e-03, -1.20885060e-03,  6.35773146e+00,
       -1.32037297e-01, -7.74407841e-02, -2.51188085e-03])

The quatitative results from my model is the array above. The feature that seems the most important is the eighth feature which is density because it has the highest beta_hat. The magnitude of the features do have a significant impact on the results. The feautres that has high magnitude, fixed acidity, free sulfur dioxide, total sulfur dioxide, alcohol, etc, all have a significantly lower beta_hat.

In [14]:
train_data_fit = RSS(beta_hat,X_train,y_train)
train_data_fit

847.5991662131569

In [15]:
test_data_fit = RSS(beta_hat,X_test,y_test)
test_data_fit

189.78413868232155

The goodness of fit on the test data is 189.78.

Try different initial beta values:

In [16]:
beta0 = np.random.normal(20,5,X_train.shape[1])
res = minimize(fun=RSS, x0=beta0, args=(X_train, y_train))
beta_hat = res.x
beta_hat

array([-2.10682457e-02, -9.63404025e-02,  8.42211329e-04,  1.48925174e-03,
        2.10807926e-02,  5.78577603e-03, -1.20886216e-03,  6.35774290e+00,
       -1.32040338e-01, -7.74405918e-02, -2.51186148e-03])

In [17]:
test_data_fit = RSS(beta_hat,X_test,y_test)
test_data_fit

189.7841846335582

In [18]:
beta0 = np.random.normal(1000,90,X_train.shape[1])
res = minimize(fun=RSS, x0=beta0, args=(X_train, y_train))
beta_hat = res.x
beta_hat

array([-2.10681523e-02, -9.63404111e-02,  8.42127304e-04,  1.48925120e-03,
        2.10806074e-02,  5.78577295e-03, -1.20886043e-03,  6.35773770e+00,
       -1.32039083e-01, -7.74403515e-02, -2.51185283e-03])

In [19]:
test_data_fit = RSS(beta_hat,X_test,y_test)
test_data_fit

189.78417549215368

In [20]:
beta0 = np.random.normal(1,100,X_train.shape[1])
res = minimize(fun=RSS, x0=beta0, args=(X_train, y_train))
beta_hat = res.x
beta_hat

array([-2.10682085e-02, -9.63403624e-02,  8.42265383e-04,  1.48924969e-03,
        2.10819371e-02,  5.78577528e-03, -1.20886154e-03,  6.35774046e+00,
       -1.32039765e-01, -7.74406537e-02, -2.51185076e-03])

In [21]:
test_data_fit = RSS(beta_hat,X_test,y_test)
test_data_fit

189.7841850425548

According to the results above, no matter how you change the initial values of beta, the end result and RSS always stay the same.

Try different solvers:

In [22]:
beta0 = np.random.normal(0,1,X_train.shape[1])
res = minimize(fun=RSS, x0=beta0, args=(X_train, y_train), method = "Nelder-Mead")
beta_hat = res.x
beta_hat

array([-8.28548939e-02, -1.99602478e+00, -1.57280353e+00, -1.53050984e-02,
       -3.56051306e+00,  1.81040054e-03,  1.23629748e-04,  1.59776906e+01,
       -2.34905680e+00, -2.79287676e-01,  1.57954978e-02])

In [23]:
test_data_fit = RSS(beta_hat,X_test,y_test)
test_data_fit

237.6179310713567

In [24]:
beta0 = np.random.normal(0,1,X_train.shape[1])
res = minimize(fun=RSS, x0=beta0, args=(X_train, y_train), method = "Powell")
beta_hat = res.x
beta_hat

array([ 6.09880708e-02, -1.70481934e-01,  9.85678170e-03,  5.34514612e-03,
        9.11057847e-01,  2.99906294e-03,  1.48628481e-04,  1.22135749e+00,
        1.22811315e+00, -3.10086404e-02, -1.96883604e-02])

In [25]:
test_data_fit = RSS(beta_hat,X_test,y_test)
test_data_fit

187.96752907628104

In [26]:
beta0 = np.random.normal(0,1,X_train.shape[1])
res = minimize(fun=RSS, x0=beta0, args=(X_train, y_train), method = "BFGS")
beta_hat = res.x
beta_hat

array([-2.10681628e-02, -9.63405936e-02,  8.41965058e-04,  1.48925046e-03,
        2.10828325e-02,  5.78577361e-03, -1.20886050e-03,  6.35773866e+00,
       -1.32039350e-01, -7.74406487e-02, -2.51183767e-03])

In [27]:
test_data_fit = RSS(beta_hat,X_test,y_test)
test_data_fit

189.78418302262367

In [28]:
beta0 = np.random.normal(0,1,X_train.shape[1])
res = minimize(fun=RSS, x0=beta0, args=(X_train, y_train), method = "TNC")
beta_hat = res.x
beta_hat

array([ 0.17987782, -0.59828997, -0.93879835,  0.00423484, -0.64246185,
       -0.0033318 ,  0.00320332, -0.88697472,  1.57491826, -0.19115733,
        0.04238404])

In [29]:
test_data_fit = RSS(beta_hat,X_test,y_test)
test_data_fit

196.9638186323669

Different solver methods do change the end result and RSS but not significantly. The density still is the most important feature and RSS are all around 189.




 
  
Regularizing the Model

Ridge

In [40]:
def ridge(beta,X,y):
    lam = 0.01
    return RSS(beta,X,y)+lam*np.sum(beta**2)

beta0 = np.random.normal(0,1,X_train.shape[1])
res = minimize(fun=ridge, x0=beta0, args=(X_train, y_train))
beta_hat = res.x
beta_hat

array([-1.85004655e-02, -9.63530652e-02, -1.18526687e-03,  1.45496910e-03,
        4.80377332e-02,  5.72134375e-03, -1.16297937e-03,  6.21393268e+00,
       -9.82437934e-02, -7.61526120e-02, -1.93401817e-03])

In [41]:
train_data_fit = RSS(beta_hat,X_train,y_train)
train_data_fit

848.0034743802443

In [42]:
test_data_fit = RSS(beta_hat,X_test,y_test)
test_data_fit

190.02797816326438

Both RSS on training and testing data increased by a tiny amount.

In [43]:
l = [0.001,0.01,0.1,1,2,5,10]
RSS_list = []

for i in l:
    def ridge(beta,X,y):
        return RSS(beta,X,y)+i*np.sum(beta**2)

    beta0 = np.random.normal(0,1,X_train.shape[1])
    res = minimize(fun=ridge, x0=beta0, args=(X_train, y_train))
    beta_hat = res.x
    
    RSS_list.append(RSS(beta_hat,X_test,y_test))

In [44]:
 RSS_list

[190.09838392763464,
 190.02799222518425,
 189.46743215114796,
 188.3408692349949,
 188.52362003736047,
 189.31185718730688,
 190.47349517026484]

Among all of the lambda values I tried, the one that yields the lowest RSS is lambda = 1. RSS = 188.34. A little lower than the linear model RSS.

Lasso

In [48]:
l = [0.01,0.1,1,2]
RSS_list = []

for i in l:
    print("Lambda =", i)
    def lasso(beta,X,y):
        return RSS(beta,X,y)+i*np.sum(np.absolute(beta))

    beta0 = np.random.normal(0,1,X_train.shape[1])
    res = minimize(fun=lasso, x0=beta0, args=(X_train, y_train))
    beta_hat = res.x
    
    print("RSS =", RSS(beta_hat,X_test,y_test))
    
    

Lambda = 0.01
RSS = 189.77150106606288
Lambda = 0.1
RSS = 189.65874329055333
Lambda = 1
RSS = 189.13209631834172
Lambda = 2
RSS = 189.18537960900886


RSS is minimized when lambda = 1. This is same as ridge regularization. It is also a little bit lower than the original RSS.

I think the magnitude of the features will still affect the results with regularization but it will be diminished because regularization prevents we use some non-related features.